File: interface/tests/matlab/demo_nonlinear_elasticity.m
clear pde;
gf_workspace('clear all');

% set a custom colormap
r=[0.7 .7 .7]; l = r(end,:); s=63; s1=20; s2=25; s3=48;s4=55; for i=1:s, c1 = max(min((i-s1)/(s2-s1),1),0);c2 = max(min((i-s3)/(s4-s3),1),0); r(end+1,:)=(1-c2)*((1-c1)*l + c1*[1 0 0]) + c2*[1 .8 .2]; end; colormap(r);

new_bricks = 1; % new brick system or old one.

incompressible = 1

lawname = 'Ciarlet Geymonat';
params = [1;1;-1.4];
params = [0;1];
if (incompressible)
    lawname = 'Mooney Rivlin';
    params = [1;1];
end

if 0,
  h=20
  % import the mesh
  %m=gfMesh('load', 'ladder.mesh');
  %m=gfMesh('load', 'ladder_1500.mesh');
  m=gfMesh('load', 'holed_bar.mesh');
  set(m, 'transform', [1 0 0; 0 0 1; 0 1 0]);
  mfu=gfMeshFem(m,3);     % mesh-fem supporting a 3D-vector field
  mfd=gfMeshFem(m,1);     % scalar mesh_fem
  % the mesh_im stores the integration methods for each tetrahedron
  mim=gfMeshIm(m,gfInteg('IM_TETRAHEDRON(5)'));
  % we choose a P2 fem for the main unknown
  set(mfu, 'fem',gfFem('FEM_HERMITE(3)'));
  %set(mfu, 'fem',gfFem('FEM_PK(3,2)'));
  mfdu=gfMeshFem(m,1);
  % the material is homogeneous, hence we use a P0 fem for the data
  gf_mesh_fem_set(mfd,'fem',gf_fem('FEM_PK(3,1)'));
  % the P2 fem is not derivable across elements, hence we use a discontinuous
  % fem for the derivative of U.
  gf_mesh_fem_set(mfdu,'fem',gf_fem('FEM_PK_DISCONTINUOUS(3,2)'));
else
  N1=4; N2=16; h=20;
  m=gf_mesh('cartesian',(0:N1)/N1 - .5, (0:N2)/N2*h, ((0:N1)/N1 - .5)*3);
  mfu=gf_mesh_fem(m,3);     % mesh-fem supporting a 3D-vector field
  mfd=gf_mesh_fem(m,1);     % scalar mesh_fem
  % the mesh_im stores the integration methods for each tetrahedron
  mim=gf_mesh_im(m,gf_Integ('IM_GAUSS_PARALLELEPIPED(3,6)'));
  % we choose a P2 fem for the main unknown
  gf_mesh_fem_set(mfu, 'fem',gf_Fem('FEM_QK(3,2)'));
  mfdu=gf_mesh_fem(m,1);
  % the material is homogeneous, hence we use a P0 fem for the data
  gf_mesh_fem_set(mfd,'fem',gf_fem('FEM_QK(3,1)'));
  % the P2 fem is not derivable across elements, hence we use a discontinuous
  % fem for the derivative of U.
  gf_mesh_fem_set(mfdu,'fem',gf_fem('FEM_QK_DISCONTINUOUS(3,2)'));
end;

m_char=gf_mesh_get(m, 'char');
mfu_char=gf_mesh_fem_get(mfu, 'char');
mfdu_char=gf_mesh_fem_get(mfdu, 'char');

% display some informations about the mesh
disp(sprintf('nbcvs=%d, nbpts=%d, nbdof=%d',gf_mesh_get(m,'nbcvs'),...
            gf_mesh_get(m,'nbpts'),gf_mesh_fem_get(mfu,'nbdof')));
P=gf_mesh_get(m,'pts'); % get list of mesh points coordinates
%pidtop=find(abs(P(2,:)-13)<1e-6); % find those on top of the object
%pidbot=find(abs(P(2,:)+10)<1e-6); % find those on the bottom

pidtop=find(abs(P(2,:)-h)<1e-6); % find those on top of the object
pidbot=find(abs(P(2,:)-0)<1e-6); % find those on the bottom


% build the list of faces from the list of points
ftop=gf_mesh_get(m,'faces from pid',pidtop); 
fbot=gf_mesh_get(m,'faces from pid',pidbot);
% assign boundary numbers
gf_mesh_set(m,'boundary',1,ftop);
gf_mesh_set(m,'boundary',2,fbot);
gf_mesh_set(m,'boundary',3,[ftop fbot]);



if (new_bricks)
  md=gf_model('real');
  gf_model_set(md, 'add fem variable', 'u', mfu);
  gf_model_set(md,'add initialized data','params', params);
  gf_model_set(md, 'add nonlinear elasticity brick', mim, 'u', lawname, 'params');
  if (incompressible)
    mfp = gf_Mesh_Fem(m,1); 
    gf_mesh_fem_set(mfp, 'classical discontinuous fem', 1);
    gf_model_set(md, 'add fem variable', 'p', mfp);
    gf_model_set(md, 'add nonlinear incompressibility brick',  mim, 'u', 'p')
  end
 
  gf_model_set(md, 'add fem data', 'DirichletData', mfd, 3);
  gf_model_set(md, 'add Dirichlet condition with penalization', mim, 'u', 1e10, 3, 'DirichletData');
  
else
  if ~incompressible,
    b0=gfMdBrick('nonlinear_elasticity', mim, mfu, lawname);
    b1=b0;
    set(b1, 'param', 'params', params);
  else
    b0=gf_MdBrick('nonlinear_elasticity', mim, mfu, lawname);
    mfp = gf_Mesh_Fem(m,1); 
    gf_mesh_fem_set(mfp, 'classical discontinuous fem', 1);
    b1=gf_MdBrick('nonlinear_elasticity_incompressibility_term',b0,mfp);
  end
  %b2=gfMdBrick('dirichlet', b1, 2);
  b3=gf_MdBrick('dirichlet', b1, 3, mfu, 'penalized');

  mds=gf_MdState(b3)
end

VM=zeros(1,gf_mesh_fem_get(mfdu,'nbdof'));

reload = 0;

if (reload == 0),
  UU=[];
  VVM=[];
  nbstep=40
else
  load 'demo_nonlinear_elasticity_U.mat';
  nb_step = size(UU,1);
end;
P=gf_mesh_fem_get(mfd, 'basic dof_nodes');
r = sqrt(P(1 ,:).^2 + P(3, :).^2);
theta = atan2(P(3,:),P(1,:));

for step=1:nbstep,
  w = 3*step/nbstep;
  %set(b2, 'param', 'R', [0;0;0]);

  if (~reload),
    R=zeros(3, gf_mesh_fem_get(mfd, 'nbdof'));
    dtheta =  pi;
    dtheta2 = pi/2;
    
    i_top = gf_mesh_fem_get(mfd, 'basic dof on region', 1);
    i_bot = gf_mesh_fem_get(mfd, 'basic dof on region', 2);
    dd = max(P(1,i_top)*sin(w*dtheta));
    if (w < 1), 
      RT1 = axrot_matrix([0 h*.75 0], [0 h*.75 1], w*dtheta);
      RT2 = axrot_matrix([0 0 0], [0 1 0], sqrt(w)*dtheta2);
      RB1 = axrot_matrix([0 h*.25 0], [0 h*.25 1], -w*dtheta);
      RB2 = RT2';
    elseif (w < 2),
      RT1 = axrot_matrix([0 h*.75 0], [0 h*.75 1], (2-w)*dtheta);
      RT2 = axrot_matrix([0 0 0], [0 1 0], w*dtheta2);
      RB1 = axrot_matrix([0 h*.25 0], [0 h*.25 1], -(2-w)*dtheta);
      RB2 = RT2';
    else
      RT1 = axrot_matrix([0 h*.75 0], [0 h*.75 1], 0);
      RT2 = axrot_matrix([0 0 0], [0 1 0], (3-w)*2*dtheta2);
      RB1 = axrot_matrix([0 h*.25 0], [0 h*.25 1], 0);
      RB2 = RT2';    
    end;
    for i=i_top,
      ro = RT1*RT2*[P(:,i);1];
      R(:, i) = ro(1:3) - P(:,i);
    end
    for i=i_bot,
      ro = RB1*RB2*[P(:,i);1];
      R(:, i) = ro(1:3) - P(:,i);
    end
    if (new_bricks)
      gf_model_set(md, 'variable', 'DirichletData', R);
      gf_model_get(md, 'solve', 'very noisy', 'max_iter', 100, 'max_res', 1e-5, 'lsearch', 'simplest');
      % full(gf_model_get(md, 'tangent matrix'))
      U = gf_model_get(md, 'variable', 'u');
      VM = gf_model_get(md, 'compute Von Mises or Tresca', 'u', lawname, 'params', mfdu);
      % sigma = gf_model_get(md, 'compute second Piola Kirchhoff tensor', 'u', lawname, 'params', mfdu);
      % sigma
    else
      gf_MdBrick_set(b3, 'param', 'R', mfd, R);
      gf_MdBrick_get(b3, 'solve', mds, 'very noisy', 'max_iter', 100, 'max_res', 1e-5);
      % full(gf_MdState_get(mds, 'tangent matrix'))
      U=gf_MdState_get(mds, 'state'); U=U(1:gf_mesh_fem_get(mfu, 'nbdof'));
      VM = gf_MdBrick_get(b0, 'von mises', mds, mfdu);
    end
    UU = [UU;U]; 
    VVM = [VVM;VM];
    save demo_nonlinear_elasticity_U.mat UU VVM m_char mfu_char mfdu_char;
  else
    U=UU(step,:);
    VM=VVM(step,:);
  end;
  disp(sprintf('step %d/%d : |U| = %g',step,nbstep,norm(U)));

  gf_plot(mfdu,VM,'mesh','off', 'cvlst',gf_mesh_get(mfdu,'outer faces'), 'deformation',U,'deformation_mf',mfu,'deformation_scale', 1, 'refine', 8); colorbar;
  axis([-3     6     0    20    -2     2]); caxis([0 .3]);
  view(30+20*w, 23+30*w);  
  campos([50 -30 80]);
  camva(8);
  camup
  camlight; 
  axis off;
  pause(1); 
  % save a picture..
  %print(gcf, '-dpng', '-r150', sprintf('torsion%03d',step));
end;
  
disp('end of computations, you can now replay the animation with')
disp('demo_nonlinear_elasticity_anim')

File: interface/tests/python/demo_nonlinear_elasticity.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Python GetFEM interface
#
# Copyright (C) 2012-2020 Yves Renard.
#
# This file is a part of GetFEM
#
# GetFEM  is  free software;  you  can  redistribute  it  and/or modify it
# under  the  terms  of the  GNU  Lesser General Public License as published
# by  the  Free Software Foundation;  either version 3 of the License,  or
# (at your option) any later version along with the GCC Runtime Library
# Exception either version 3.1 or (at your optifn) any later version.
# This program  is  distributed  in  the  hope  that it will be useful,  but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or  FITNESS  FOR  A PARTICULAR PURPOSE.  See the GNU Lesser General Public
# License and GCC Runtime Library Exception for more details.
# You  should  have received a copy of the GNU Lesser General Public License
# along  with  this program;  if not, write to the Free Software Foundation,
# Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301, USA.
#
############################################################################

import numpy as np
from numpy import linalg as npla

import getfem as gf

gf.util_trace_level(1)

dirichlet_version = 2          # 1 = simplification, 2 = penalisation
test_tangent_matrix = False    # Test or not tangent system validity
incompressible = False         # Incompressibility option
explicit_potential = False     # Elasticity law with explicit potential

# lawname = 'Ciarlet Geymonat'
# params = [1.,1.,0.25]
# lawname = 'SaintVenant Kirchhoff'
# params = [1.,1.]
lawname = 'Compressible Mooney Rivlin'
params = [1.,1.,2.]
if (incompressible):
    lawname = 'Incompressible Mooney Rivlin'
    params = [1.,1.]

N1 = 2; N2 = 4; h = 20.; DX = 1./N1; DY = (1.*h)/N2;
m = gf.Mesh('cartesian', np.arange(-0.5, 0.5+DX,DX), np.arange(0., h+DY,DY),
            np.arange(-1.5, 1.5+3*DX,3*DX))
mfu  = gf.MeshFem(m, 3)            # mesh-fem supporting a 3D-vector field
mfdu = gf.MeshFem(m,1)
# The mesh_im stores the integration methods for each tetrahedron
mim = gf.MeshIm(m, gf.Integ('IM_GAUSS_PARALLELEPIPED(3,4)'))
# We choose a P2 fem for the main unknown
mfu.set_fem(gf.Fem('FEM_QK(3,2)'))

if (dirichlet_version == 1):
  mfd = mfu;
else:
  mfd = gf.MeshFem(m,1)
  mfd.set_fem(gf.Fem('FEM_QK(3,1)'))

# The P2 fem is not derivable across elements, hence we use a discontinuous
# fem for the derivative of U.
mfdu.set_fem(gf.Fem('FEM_QK_DISCONTINUOUS(3,2)'));

# Display some information about the mesh
print('nbcvs=%d, nbpts=%d, nbdof=%d' % (m.nbcvs(), m.nbpts(), mfu.nbdof()))

# Assign boundary numbers

ftop = m.outer_faces_with_direction([0.,  1., 0.], 0.5)
fbot = m.outer_faces_with_direction([0., -1., 0.], 0.5)

m.set_region(1, ftop);
m.set_region(2, fbot);
m.set_region(3, np.append(ftop,fbot,axis=1));

# Model definition

md=gf.Model('real')
md.add_fem_variable('u', mfu)
md.add_initialized_data('params', params)

if (not(explicit_potential)):
    md.add_finite_strain_elasticity_brick(mim, lawname, 'u', 'params')
else:
    print("Explicit elastic potential")
    K = 1.2; mu = 3.0;
    md.add_macro('F_',  '(Id(meshdim)+Grad_u)')
    md.add_macro('J_',  'Det(F_)')
    md.add_macro('be_', 'Left_Cauchy_Green(F_)')
    md.add_initialized_data('K',  [K])
    md.add_initialized_data('mu', [mu])
    md.add_initialized_data('paramsIMR', [1,1,2])
    _expr_1 = "(K/2)*sqr(log(J_))+(mu/2)*(Matrix_j1(be_)-3)";
    _expr_2 = "(K/2)*sqr(log(J_))+(mu/2)*(pow(Det(be_),-1./3.)*Trace(be_)-3)"
    _expr_3 = "paramsIMR(1)*(Matrix_j1(Right_Cauchy_Green(F_))-3)+ paramsIMR(2)*(Matrix_j2(Right_Cauchy_Green(F_)) - 3)+paramsIMR(3)*sqr(J_-1)"
    md.add_nonlinear_term(mim, _expr_3);

# md.add_nonlinear_term(mim, 'sqr(Trace(Green_Lagrangian(Id(meshdim)+Grad_u)))/8 + Norm_sqr(Green_Lagrangian(Id(meshdim)+Grad_u))/4')
# md.add_nonlinear_term(mim, '((Id(meshdim)+Grad_u)*(params(1)*Trace(Green_Lagrangian(Id(meshdim)+Grad_u))*Id(meshdim)+2*params(2)*Green_Lagrangian(Id(meshdim)+Grad_u))):Grad_Test_u')
# md.add_nonlinear_term(mim, 'Saint_Venant_Kirchhoff_potential(Grad_u,params)')
    
if (incompressible):
    mfp = gf.MeshFem(m,1)
    mfp.set_classical_discontinuous_fem(1)
    md.add_fem_variable('p', mfp)
    md.add_finite_strain_incompressibility_brick(mim, 'u', 'p')
    # md.add_nonlinear_term(mim, 'p*(1-Det(Id(meshdim)+Grad_u))')
    # md.add_nonlinear_term(mim, '-p*Det(Id(meshdim)+Grad_u)*(Inv(Id(meshdim)+Grad_u))'':Grad_Test_u + Test_p*(1-Det(Id(meshdim)+Grad_u))')


if (dirichlet_version == 1):
    md.add_fem_data('DirichletData', mfu)
    md.add_Dirichlet_condition_with_simplification('u', 3, 'DirichletData')
else:
    md.add_fem_data('DirichletData', mfd, 3)
    md.add_Dirichlet_condition_with_penalization(mim, 'u', 1e4, 3, 'DirichletData')


VM=np.zeros(mfdu.nbdof());
UU=np.zeros(0);
VVM=np.zeros(0);
nbstep=40;


P = mfd.basic_dof_nodes()
r = np.sqrt(np.square(P[0 ,:]) + np.square(P[2, :]))
theta = np.arctan2(P[2,:], P[0,:]);


def axrot_matrix(A, B, theta):
    n=(np.array(B)-np.array(A)); n = n/npla.norm(n)
    a=n[0]; b=n[1]; c=n[2]
    d=np.sqrt(b*b+c*c)
    T=np.eye(4); T[0:3,3]=-np.array(A)
    Rx=np.eye(4)
    if (npla.norm(n[1:3])>1e-6):
        Rx[1:3,1:3]=np.array([[c/d,-b/d],[b/d,c/d]])
    Ry=np.eye(4); Ry[[[0],[2]],[0,2]]=[[d,-a],[a,d]]
    Rz=np.eye(4)
    Rz[0:2,0:2]=np.array([[np.cos(theta),np.sin(theta)],
                          [-np.sin(theta),np.cos(theta)]])
    R = np.dot(np.dot(np.dot(npla.inv(T),npla.inv(Rx)),
                      np.dot(npla.inv(Ry),Rz)),np.dot(np.dot(Ry,Rx),T))
    return R;
  



for step in range(1,nbstep+1):
    w = (3.*step)/nbstep

    # Computation of the rotation for Dirichlet's condition
    dtheta =  np.pi
    dtheta2 = np.pi/2.
      
    if (dirichlet_version == 1):
        R=np.zeros(mfd.nbdof())
    else:
        R=np.zeros((3, mfd.nbdof()))
    
    i_top = mfd.basic_dof_on_region(1)
    i_bot = mfd.basic_dof_on_region(2)
    
    dd = np.amax(P[1,i_top]*np.sin(w*dtheta))
    
    if (w < 1): 
        RT1 = axrot_matrix([0, h*.75, 0], [0, h*.75, 1], w*dtheta)
        RT2 = axrot_matrix([0, 0, 0], [0, 1, 0], np.sqrt(w)*dtheta2)
        RB1 = axrot_matrix([0, h*.25, 0], [0, h*.25, 1], -w*dtheta)
        RB2 = RT2.transpose()
    elif (w < 2):
        RT1 = axrot_matrix([0, h*.75, 0], [0, h*.75, 1], (2-w)*dtheta);
        RT2 = axrot_matrix([0, 0, 0], [0, 1, 0], w*dtheta2);
        RB1 = axrot_matrix([0, h*.25, 0], [0, h*.25, 1], -(2-w)*dtheta);
        RB2 = RT2.transpose()
    else:
        RT1 = axrot_matrix([0, h*.75, 0], [0, h*.75, 1], 0);
        RT2 = axrot_matrix([0, 0, 0], [0, 1, 0], (3-w)*2*dtheta2);
        RB1 = axrot_matrix([0, h*.25, 0], [0, h*.25, 1], 0);
        RB2 = RT2.transpose()

    if (dirichlet_version == 1):
        for i in i_top:
            ro = np.dot(RT1, np.dot(RT2,np.append(P[:,i],[1])))
            R[i] = ro[mod(i,3)] - P[mod(i,3),i]
        for i in i_bot:
            ro = np.dot(RB1,np.dot(RB2,np.append(P[:,i],[1])))
            R[i] = ro[mod(i,3)] - P[mod(i,3),i] 
    else:
        for i in i_top:
            ro = np.dot(RT1,np.dot(RT2,np.append(P[:,i],[1])))
            R[:, i] = ro[0:3] - P[:,i]
        for i in i_bot:
            ro = np.dot(RB1,np.dot(RB2,np.append(P[:,i],[1])))
            R[:, i] = ro[0:3] - P[:,i]
    
    md.set_variable('DirichletData', R)

    # Nonlinear solve
    md.solve('very noisy', 'max_iter', 50, 'max_res', 1e-7, 'lsearch',
             'simplest')

    print("Iteration %d done" % step)

    if (test_tangent_matrix):
       md.test_tangent_matrix(1E-8, 10, 0.0001)

    U = md.variable('u')

    VM0 = md.compute_Von_Mises_or_Tresca('u', lawname, 'params', mfdu)

    # Direct interpolation of the Von Mises stress
    # VM = md.interpolation('(sqrt(3/2)/Det(Id(meshdim)+Grad_u))*Norm((Id(meshdim)+Grad_u)*Saint_Venant_Kirchhoff_PK2(Grad_u,params)*(Id(meshdim)+Grad_u'') - Id(meshdim)*Trace((Id(meshdim)+Grad_u)*Saint_Venant_Kirchhoff_PK2(Grad_u,params)*(Id(meshdim)+Grad_u''))/meshdim)', mfdu);

    VM = md.compute_finite_strain_elasticity_Von_Mises(lawname, 'u', 'params', mfdu)
    print(npla.norm(VM-VM0))
    
    sl=gf.Slice(('boundary',), mfu, 4)
    sl.export_to_vtk('demo_nonlinear_elasticity_iter_%d.vtk' % step, 'ascii',
                     mfdu,  VM, 'Von Mises Stress', mfu, U, 'Displacement')



print('You can vizualize the loading steps by launching for instance')
print('mayavi2 -d demo_nonlinear_elasticity_iter_1.vtk -f WarpVector -m Surface')
